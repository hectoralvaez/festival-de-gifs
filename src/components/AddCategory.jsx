import { useState } from "react"

export const AddCategory = () => {
    const [inputValue, setInputValue] = useState('One Punch');

    // const handleInputChange = (event) => {  // De esta manera tendríamos que usar "event.target.value"
    const handleInputChange = ( {target} ) => {  // Si desestructuramos el objeto y solo nos interesa el "target", se puede usar directamente "target.value"
        // console.log(event);     // Con esto imprimimos toda la información del objeto "event"
        // console.log(target.value);         
        // setInputValue('Hola Mundo');    // De esta manera, en cuanto tocamos una tecla, aparece "Hola Mundo", y no se puede cambiar
        setInputValue( target.value );
    }

    const handleSubmit = ( event ) => {
        event.preventDefault();     // Con event.preventDefault(), evitamos que se recargue la página (refresh del navegador web) por el event "submit"
        console.log(inputValue);
    }

  return (
    <form onSubmit={ handleSubmit }>    {/* Con el uso del formulario no aseguramos que al ENVIARLO (con botón o con la tecla enter) hace la
                                            acción de enviar la información que hay en el imput.
                                            Si no se hiciera así, habría que controlar que cuando está escribiendo en el input, al darle a la tecla "enter"
                                            envie el contenido */}
        <input
            // Eso parecen atributos html, pero en realidad son "properties" (props) de React
            type="text"
            placeholder="Buscar gifs"
            value={inputValue}
            // onChange={ handleInputChange }    // Para cambiar el contenido hay que aplicar el metodo "onChange" 
            // onChange={ ( event ) => handleInputChange(event) }      //  Al estar dentro de la "prop" "onChange", lo que está haciendo es 
                                                                    // recibir el evento "React.ChangeEvent" y ese evento enviarlo a la función "handleInputChange"
    
            // onChange={ ( event ) => handleInputChange(event) }   // En este caso, como estamos mandando una función cuyo primer argumento es el argumento que estoy 
                                                                    // mandándole a la función que quiero ejecutar, podemos obviar poner la función de flecha, ya que 
                                                                    // solo envía el "event" y solo con enviar la referencia a la función es suficiente, ya que 
                                                                    // siempre va a recibir el "event"
            // TOTAL, que queda así:
            onChange={ handleInputChange }
        />
    </form>
  )
}