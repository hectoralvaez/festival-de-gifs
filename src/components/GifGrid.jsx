import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifItem } from "./GifItem";

export const GifGrid = ({ category }) => {
    // Creamos un estado local para mantener las imágenes dentro del "GifGrid"
    // const [images, setImages] = useState([]); // Inicializamos el estado con un Array vacío, podría ser "false", pero al dejarlo vacío, luego controlamos cuando el array tenga contenido para mostrarlo.

    // NO LLAMAR NUNCA LA EJECUCIÓN DE UNA FUNCIÓN dentro de un functional component
    // Cada vez que se llama al Functional Component <GifGrid /> y se renderiza, vuelve a ejecutar la función getGifs(category)
    // getGifs(category);  // PENDIENTE: mover esta llamada a la función fuera del functional component para que no se dispare cada vez que se llama al functional component.
    // useEffect( () => {
    //     getGifs(category);  // RESUELTO: Dentro del useEffect, ahora solo se dispara cada vez que se llama al functional component "GifGrid".
    // }, [ ]);    // Si no se definen las dependencias en el array, significa que simplemente solo se v a ejecutar la primera vez 

    // ESTO NO ES CORRECTO:
    // useEffect no puede usar "async", tiene que retornar una función, no una promesa
    // useEffect ( async() => {
    //     const newImages = await getGifs(category);
    //     setImages(newImages);
    // }, []);
    
    // CORRECTO VERSIÓN 1:
    // useEffect ( () => {
    //     getGifs(category)
    //     .then( newImages => setImages(newImages));
    // }, []);
    // FIN CORRECTO VERSIÓN 1:

    // CORRECTO VERSIÓN 2:
    // Esta función SÍ puede ser asíncrona, podemos pasarle la promesa y luego llamar a esta función dentro del "useEffect" 
    // para que no se ejecute siempre, solo la primera vez que carga el componente
    // const getImages = async() => {
    //     const newImages = await getGifs( category );
    //     setImages(newImages);
    // }

    // useEffect ( () => {
    //     getImages();
    // }, []);
    // FIN CORRECTO VERSIÓN 2:

    const { images, isLoading } = useFetchGifs ( category );
    return (
        <>
            <h3>{ category }</h3>
            <div className="card-grid">
                {
                    // VERSIÓN 1 (desestructurando el objeto para sacar las props que necesitamos de manera independiente)
                    // images.map( ({id, title, url}) => (      // usamos el "map" de "images" que es donde se está guardando el useState de "setImages"
                    //     <GifItem 
                    //         key={id} 
                    //         title={title}
                    //         url={url}
                    //     />
                    //  ) ) 

                    // VERSIÓN 2 (pasando como props "image" para enviar todas las propiedasdes completas)
                    // images.map( ( image ) => (      // usamos el "map" de "images" que es donde se está guardando el useState de "setImages"
                    //     <GifItem 
                    //         key={image.id} 
                    //         title={image.title}
                    //         url={image.url}
                    //     />
                    //  ) ) 
                    // VERSIÓN 3 (esparcir las "props", de esta manera el component recibe TODAS las properties)
                    images.map( ( image ) => (      // usamos el "map" de "images" que es donde se está guardando el useState de "setImages"
                        <GifItem 
                            key={image.id} 
                            { ...image }
                        />
                     ) ) 
                }
            </div>
        </>
    )
}
