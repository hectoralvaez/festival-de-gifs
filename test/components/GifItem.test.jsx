import { render } from "@testing-library/react"
import { GifItem } from "../../src/components/GifItem";


describe("Pruebas en <GifItem />", () => {
    const title = 'el tÃ­tulo';
    const url = 'http://www.google.com';

    test("Tiene que hacer match con el snapshot", () => {
        const { container } = render(<GifItem title={title} url={url} />);
        expect(container).toMatchSnapshot();
    });
});