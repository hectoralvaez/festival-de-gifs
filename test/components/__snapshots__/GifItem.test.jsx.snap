// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> Tiene que hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="el título"
      src="http://www.google.com"
    />
    <p>
      el título
    </p>
  </div>
</div>
`;
